Add chapter on Optuna for Hyperparameter Optimization (#138)
allenai/allennlp-guide: chapters/part3/hyperparameter-optimization-with-optuna.md

---
title: 'Hyperparameter Optimization with Optuna'
description: "This chapter gives a basic tutorial for optimizing the hyperparameters of your model."

> type: chapter

Reviewer: You can put an "author" key in here, to have you get credit for writing this (won't render correctly until #134 is merged, though).

Author: Added in 1119a64. Thank you.

---

<textblock>

In this chapter, we'll give a quick tutorial of hyperparameter optimization for your AllenNLP model.

</textblock>

> 

Author: :+1:

        trial=trial,  # trial object
        config_file="./config/imdb_optuna.jsonnet",  # path to jsonnet
        serialization_dir=f"./result/optuna/{trial.number}",
        metrics="best_validation_accuracy"
    )
    return executor.run()
```

After defining search spaces using `trial.suggest_int` and `trial.suggest_float`, the `trial` object should be passed to `AllenNLPExecutor`.
The `trial` object holds all suggested values after defining search spaces using `trial.suggest_int` and `trial.suggest_float`,  and it should be pass to `AllenNLPExecutor`.
`AllenNLPExecutor` takes four required arguments; `trial` (Optuna's object), `config_file` (path to a model configuration),
`serialization_dir` (directory for saving model snapshot, log, etc.), and `metrics` you want to optimize.
In the above example, we create an instance of `AllenNLPExecutor` as `executor`.
Once the `executor` instance is created, training is started with `executor.run()`.
In each trial step in optimization, the objective function is called and does the following steps:

1. Train a model (`executor.run()`)
2. Return a target metric on validation data (`executor.run()` returns the specified metric)

> Now, we finished defining objective function.
> Let's write Optuna stuff for launching optimization!

Reviewer: ```suggestion
Once we've written an objective function, we can write a script for launching optimization.
```

In Optuna, we create a study object and pass the objective function to the `optimize()` method as follows.

> You can specify something; a way to save a result of optimization, a sampler for searching hyperparameters (`TPESampler` is based on Bayesian Optimization),
> direction for optimizing (maximize or minimize), number of jobs for distributed training, or timeout.

Reviewer: ```suggestion
You can specify something a number of parameters for the hyperparameter optimization process:
- a way to save a result of optimization
- a sampler for searching hyperparameters (`TPESampler` is based on Bayesian Optimization)
- direction for optimizing (maximize or minimize)
- number of jobs for distributed training
- timeout
and more.  An example launch script is below.
```

```python
if __name__ == '__main__':
    study = optuna.create_study(
        storage="sqlite:///result/trial.db",  # save results in DB
        sampler=optuna.samplers.TPESampler(seed=24),
        study_name="optuna_allennlp",
        direction="maximize",
    )

    timeout = 60 * 60 * 10  # timeout (sec): 60*60*10 sec => 10 hours
    study.optimize(
        objective,
        n_jobs=1,  # number of processes in parallel execution
        n_trials=30,  # number of trials to train a model
        timeout=timeout,  # threshold for executing time (sec)
    )
```

</exercise>

<exercise id="3" title="Results of Hyperparameter Optimization">

> If `study.optimize` successfully runs, `trial.db` would be created in the directory `result`.

Reviewer: If the above section is split into two, this one can probably be combined with the second one, to have one exercise talking about running optuna and seeing the results.

> [Tip] You can change a names of database file and directory by changing the value of `storage` in `optuna.create_study`.

Reviewer: This tip may not be necessary if this text immediately follows the example launch script above, where we already mentioned that we can vary the storage location.

Author: Thank you, I removed the tip since [your suggestion](https://github.com/allenai/allennlp-guide/pull/138#discussion_r472491586) explained this well. b690276

You can see and analyze a result by passing `study` object to various methods implemented in Optuna.
If you want to separate an analysis from optimization, you can save the `study` (e.g. RDB) and load it in another script.

```python
study = optuna.load_study(
  storage="sqlite:///result/trial.db",
  study_name="optuna_allennlp"
)
```

Let's check the results of trials with a `pandas` dataframe.

```python
study.trials_dataframe()
```

<img src="/part3/hyperparameter-optimization-with-optuna/trials_dataframe.jpg" alt="Dataframe">

Next, visualize a history of optimization.
To plot a history of optimization, we can use `optuna.visualization.plot_optimization_history`.

> I also put a validation accuracy of baseline model as a reference.

Reviewer: ```suggestion
We also put the validation accuracy of a baseline model as a reference.
```

It's not clear to me how this was done.

It shows that Optuna successfully found hyperparameters to achieve better performance.
Note that this figure shows one result of optimization.

> I performed optimization five times with different random seeds and got the average validation accuracy of 0.909 (±0.002), which outperforms the baseline by a large margin.

Reviewer: ```suggestion
We performed optimization five times with different random seeds and got an average validation accuracy of 0.909 (±0.002), which outperforms the baseline by a large margin.
```

```python
optuna.visualization.plot_optimization_history(study)
```

<br>
<img src="/part3/hyperparameter-optimization-with-optuna/optimization_history.jpg" alt="Plot Optimization History" />

> From the recent release of Optuna, we can evaluate parameter importances based on finished trials.

Reviewer: ```suggestion
Optuna also lets you evaluate parameter importances based on finished trials.
```

You can use `AllenNLPPruningCallback` that is the new feature of Optuna, which allows users to prune unpromising trials with algorithms implemented in Optuna.
`AllenNLPPruningCallback` is the interface to provide the way to use these pruning algorithms in `GradientDescentTrainer`, which is the standard way to train a model in AllenNLP.
To enable a pruning, you must create the AllenNLP callback first. `AllenNLPPruningCallback` needs two arguments: a trial and a target metric.
In this example, we use `validation_accuracy` for the metric to determine if a trial should be pruned or not.
Note that `patience` is set to `None` since pruning and built-in early stopping could conflict.
The objective function looks like following:

```python
    def objective(trial):
        trainer = GradientDescentTrainer(
            model=model,
            optimizer=optimizer,
            data_loader=data_loader,
            validation_data_loader=validation_data_loader,
            validation_metric="+accuracy",
            patience=None,
            num_epochs=50,
            cuda_device=CUDA_DEVICE,
            serialization_dir="result",

>             epoch_callbacks=[AllenNLPPruningCallback(trial, "validation_accuracy")],

Reviewer: This is interesting, I didn't know about this.  Is there a way to get the callback instantiated through a config file?  What would have to happen to make that work?  It's pretty annoying to have to instantiate the trainer yourself here, because it messes with a bunch of our typical recommendations for allennlp usage.  If we can make this easier, that would be better.

Reviewer: We have a mechanism for passing in arbitrary, in-memory objects when using `FromParams` (see [here](https://guide.allennlp.org/using-config-files#4)).  The main problem is that you need a line somewhere in your code that passes in a `trial` object when you call `from_params`, and you need that value to be passed down to all of your objects recursively.  I tried to figure out a way to do that with minimal modification to our train command, and there isn't a super easy way.  But, here's a bit of a hack that you can do in your own code that abuses the `FromParams` code a bit, but I think it might actually work.

You have [a method](https://github.com/optuna/optuna/blob/9b0081a212cba04e97a28ca510e5e72559b01a0d/optuna/integration/allennlp.py#L144-L145) where you're instantiating a `Params` object.  If that object had a `trial` key with the `Trial` object as a value, it will get passed to `TrainModel.from_params`.  The trouble is that it won't be treated as an `extra`, which it needs to be in order to be passed down recursively to all child classes (like the callbacks that you want to construct).  So, what you can do is create your own subclass of `TrainModel`, registered with a separate name, and have that hijack the `from_params` method and turn the `trial` key from something in the `Params` dictionary to an `extra`.  It would look something like this:

```python
@TrainModel.register("optuna")
class TrainModelWithOptuna(TrainModel):
    def from_params(# match the arguments to `FromParams.from_params`):
        # This is a pretty crazy hack of how FromParams was
        # intended to be used, but it should work for your use case
        trial = params.pop("trial")
        return TrainModel.from_params(params=params, trial=trial, **extras)
```

When objects are passed as extra `kwargs` like this, they are made available to the constructors of every object that gets constructed underneath this object in the code, regardless of whether intermediate classes need that argument (this is how the `Embedding` has access to the `Vocabulary` inside of a `TextFieldEmbedder`, even though the `TextFieldEmbedder` doesn't take a `Vocabulary` as an argument).  So, if we can get `TrainModel.from_params` to get called with `trial=trial`, then the callbacks that you have should get passed the `trial` object correctly.

In order to make this work, where you [instantiate the `Params` object](https://github.com/optuna/optuna/blob/9b0081a212cba04e97a28ca510e5e72559b01a0d/optuna/integration/allennlp.py#L144-L145) you need to add a `"type": "optuna"` key, and a `"trial": trial` key.  The `FromParams` code will call `TrainModel.from_params` with your hacked parameter object, and it will see the type is `optuna`, and call `TrainModelWithOptuna.from_params` after _popping the `type` key_.  Then inside `TrainModelWithOptuna.from_params`, we pop the `trial` key and make it an `extra` when we call `TrainModel.from_params` again.  This time, there is no `type` key, so the default `TrainModel` constructor gets called, but it now has your extra argument.

That should be all that's necessary - only one extra class and a couple of lines of modification to your `run` logic.  Understanding why this works is complicated, though, so I'd definitely recommend having lots of comments in your code explaining what's going on, if you choose to do this.

Author: Sorry for the late reply.
I found the way that enables `AllenNLPExecutor` to use `AllenNLPPruningCallback`.

We are working on adding support for Optuna pruning API in https://github.com/optuna/optuna/pull/1772.
After merging this PR, we can easily enable pruning by specifying an epoch callback.
(e.g. https://github.com/himkt/optuna-allennlp/pull/9/files)

1cb7a7a  I updated the content based on the PR.
